api/src/middleware/rate-limiter.js

```diff
@@ -0,0 +1,42 @@
+const auth = require('../auth');
+const rateLimitService = require('../services/rate-limit');
+const serverUtils = require('../server-utils');
+
+const getBasicAuthUsername = (req) => auth.basicAuthCredentials(req);
+
+const consume = (req, res) => {
+  res.on('finish', () => {
+    if (res.statusCode === 401 || res.statusCode === 429) {
+      rateLimitService.consume(getKeys(req));
```

Comment:
Only consume after an unsuccessful login attempt. This won't protect against DoS style attacks, the idea is to protect the user's password. DoS would require a much more sophisticated process...